mattermost-plugin/server/notifications.go

```diff
@@ -0,0 +1,58 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/mattermost/focalboard/server/services/notify"
+	"github.com/mattermost/focalboard/server/services/notify/notifymentions"
+
+	pluginapi "github.com/mattermost/mattermost-plugin-api"
+
+	"github.com/mattermost/mattermost-server/v6/model"
+
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+const (
+	botUsername    = "boards"
+	botDisplayname = "Boards"
+	botDescription = "Created by Boards plugin."
+)
+
+func createMentionsNotifyBackend(client *pluginapi.Client, serverRoot string, logger *mlog.Logger) (notify.Backend, error) {
+	bot := &model.Bot{
+		Username:    botUsername,
+		DisplayName: botDisplayname,
+		Description: botDescription,
+	}
+	botID, err := client.Bot.EnsureBot(bot)
+	if err != nil {
+		return nil, fmt.Errorf("failed to ensure %s bot: %w", botDisplayname, err)
+	}
+
+	delivery := &deliveryAdapter{client: client}
+
+	backend := notifymentions.New(delivery, botID, serverRoot, logger)
+
+	return backend, nil
+}
+
+type deliveryAdapter struct {
```

Comment:
I have doubts about this interface/implementation.

I think this defines how you deliver the notification, in this case through the API, but it is also leaking some implementation details related to how you resolves the destination.

My expectation would be that receives a notification to deliver, and you deliver it. For example, if I add an "email adapter" later on, why that implementation requires the `GetDirectChannel`?

Maybe our plugin based delivery adapter should receive the `GetDirect`, `CreatePost`, `GetUserByID` and `GetUserByUsername` functions at construction time, and expose `Send(changes)` method that does all the magic needed to send through the API.

Reply:
I think what you are describing is how it is implemented.  A `backend` exposes a generic method that gets called with changes and at construction time accepts the `GetDirect`, etc, functions.  This allows us to easily create different types of backends (not just for mentions). The `Delivery` interface is just for the mentions backend... other backend types (e.g. card subscription) can have different delivery interfaces.

Reply:
BTW, the only reason there is an adapter is because the new, refactored, plugin API cannot be passed as is and fulfil an interface.  All the methods have been grouped into structs within it a client struct.

Reply:
My point is. mentions can be notified through the plugin API using a message, or through an email. I see multiple delivery backends for the mentions, not only the API delivery backend, and I have the sensation that there is no other possible backend for that.

Maybe I'm understanding this all the way around. Maybe you want to have a Delivery mechanism that is for "send DMs to somebody" that can be reuse in other backends later?

Reply:
Ok, I've refactored to create a more generic `Delivery` interface, with a `PluginAPIDelivery` implementation.  We can add an `EmailDelivery` implementation, or whatever in the future.

Reply:
I really like more the current result. Thanks!